core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTextAggregator.java

```diff
@@ -0,0 +1,271 @@
+    private long lastTrieSize;
+
+    private final boolean filterDuplicateText;
+    private static final int MEMORY_GROWTH_REPORTING_INTERVAL_BYTES = 5000;
+
+
+
+    public SignificantTextAggregator(String name, AggregatorFactories factories,
+            SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators,
+            BucketCountThresholds bucketCountThresholds, IncludeExclude.StringFilter includeExclude,
+            SignificanceHeuristic significanceHeuristic, SignificantTextAggregatorFactory termsAggFactory,
+            String fieldName, boolean filterDuplicateText,
+            Map<String, Object> metaData) throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+        this.bucketCountThresholds = bucketCountThresholds;
+        this.includeExclude = includeExclude;
+        this.filterDuplicateText = filterDuplicateText;
+        this.significanceHeuristic = significanceHeuristic;
+        this.termsAggFactory = termsAggFactory;
+        this.fieldName = fieldName;
+        bucketOrds = new BytesRefHash(1, context.bigArrays());
+        dupSequenceSpotter = new DuplicateByteSequenceSpotter();        
+        lastTrieSize = dupSequenceSpotter.getEstimatedSizeInBytes();
+    }
+
+    
+    
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        final ShardGetService shardGetService = super.context.indexShard().getService();
+        final BytesRefBuilder previous = new BytesRefBuilder();
+        return new LeafBucketCollectorBase(sub, null) {
+            
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                Terms tv = ctx.reader().getTermVector(doc, fieldName);
+                if (tv!=null &&  (tv.hasPositions()  || tv.hasOffsets())){
+                    // Not necessarily faster than re-analysis in my tests!
+                    processTokenStream(doc, bucket, new TokenStreamFromTermVector(tv, -1), "");
+                    
+                } else {
+                    loadFromSource(doc, bucket, fieldName);
+                }                
+                numCollectedDocs++;
+                dupSequenceSpotter.startNewSequence();
+            }
+            
+            private void processTokenStream(int doc, long bucket, TokenStream ts, String fieldText) throws IOException{
+                if (filterDuplicateText) {
+                    ts = new DeDuplicatingTokenFilter(ts, dupSequenceSpotter);
+                }
+                CharTermAttribute termAtt = ts.addAttribute(CharTermAttribute.class);
+                ts.reset();
+                try {
+                    //Assume tokens will average 5 bytes in length to size number of tokens
+                    BytesRefHash inDocTerms = new BytesRefHash(1+(fieldText.length()/5), context.bigArrays());
```

Comment:
True. Future work is that I want to do almost highlighter type work here - translating raw indexed tokens back into something explainable to end users so looking at the original text (de-stemmed etc) and spotting runs of tokens as useful phrases rather than individual terms. 
A highlighter could perhaps also help here to pre-process and concentrate "significant text" analysis to only the sections of long documents that contained the relevant hits.

Maybe that's another case for dropping TermVectors as a source ?